website/blog/2023-03-30-relay-15.mdx

```diff
@@ -0,0 +1,153 @@
+---
+title: Relay v15.0 and VSCode Extension v2.0
+author: The Relay Team
+hide_table_of_contents: false
+---
+
+The Relay team is happy to announce the release of Relay v15 as well as v2 of the Relay VSCode extension. These releases are a long time coming and include lots of new feature, bug fixes and quality of life improvements. While this release is a major version bump and includes a couple breaking changes, few users should be affected and upgrading should be seamless for most use cases. You can find the full list of changes in the [v15 Release Notes](https://github.com/facebook/relay/releases/tag/v15.0.0).
+
+## What's new in Relay 15?
+
+### Support `@refetchable` on interfaces
+
+Previously it wasn't possible to add the `@refetchable` directive to fragments on server interface types. That support has now been added and will no longer error in the Relay compiler.
+
+```
+// schema.graphql
+
+interface RefetchableInterfaceFoo @fetchable(field_name: "id") {
+    id: ID!
+}
+
+extend type Query {
+  fetch__RefetchableInterfaceFoo(id: ID!): RefetchableInterfaceFoo
+}
+
+// fragment
+
+fragment RefetchableFragmentFoo on RefetchableInterfaceFoo
+  @refetchable(queryName: "RefetchableFragmentFooQuery") {
+  id
+}
+```
+
+### Persisted query improvements
+
+For users who are using URL-based persisted queries, you can now specify custom headers to send with the request to persist your queries. For example, this can be used to send auth headers to your query persistence URL endpoint.
+
+```js
+persistConfig: {
+  url: 'example.com/persist',
+  headers: {
+    Authorization: 'bearer TOKEN'
+  }
+}
+```
+
+And for users who are using file-based persisted queries a new feature flag, `compact_query_text` was added to remove all whitespace from persisted query output text. This can cause a substantial reduction in your persisted query text file (we've seen an example of reducing a single query text from 84kb to 31kb). This new feature flag can be enabled within your Relay config file.
+
+```js
+persistConfig: {
+  file: 'path/to/file.json',
+  algorithm: 'SHA256'
+},
+featureFlags: {
+  compact_query_text: true
+}
+```
+
+### Typesafe updates now support missing field handlers
+
+One of the few breaking changes in this release, the [missing field handlers](https://relay.dev/docs/guided-tour/reusing-cached-data/filling-in-missing-data) function signature has been changed. The `record` argument to the handler used to recieve a `Record` type (which is an untyped grabbag containing whatever the store has) but we now pass a `ReadOnlyRecordProxy` which is much more typesafe and provides a better interface. Furthermore, we also replaced the `field` argument of type `NormalizationLinkedField` with `CommonLinkedField`, which both a `ReaderLinkedField` and `NormalizationLinkedField` "implement".
```

Comment:
Not clear how the paragraph relates to the title.

Reply:
I would rewrite as:

Typesafe updaters now support missing field handlers. Previously, if you selected `node(id: 4) { ... on User { name, __typename } }` in a typesafe updater, but that user was fetched in a different way (e.g. with `best_friend { name }`), you would not be able to access and mutate that user using the typesafe updater.

In this release, we add support for missing field handlers in typesafe updaters, meaning that if a missing field handler is set up for `node` (as in the [example here](https://relay.dev/docs/next/guided-tour/reusing-cached-data/filling-in-missing-data/#internaldocs-banner)), you will be able to update the user's name with this missing field handler.

In order to support this, the signature of [missing field handlers](https://relay.dev/docs/guided-tour/reusing-cached-data/filling-in-missing-data) has been changed. The `record` argument to the handler used to recieve a `Record` type (which is an untyped grabbag of data). It now receives a `ReadOnlyRecordProxy`. Furthermore, the `field` argument of type `NormalizationLinkedField` is now `CommonLinkedField`, which is a type containing the properties found in both `ReaderLinkedField` and `NormalizationLinkedField`.

---------

Okay this isn't my best writing, but I think that explains why we did this